### ISSUE TITLE
vdf witness_type distribution changed since ASIC timelords ramp up

### ISSUE THREAD

**Issue by neurosis69** [Chia-Network/chiavdf#159]
We are just evaluating if another blueboxing run would make sense and therefore we are doing some blockchain analysis.

Since the ASIC timelords got activated (assuming it was around block height 4387596), the variance of the witness types for the different vdf proofs surprisingly increased by a lot and not only towards a more compacted proof.

For instance, the avg proof size for rc_ip (reward_chain_ip_proof) got increased from about 382B before to 712B after.

### Questions

- Is this an expected behaviour with the new ASIC timelords?
- Could this be caused by a bug in the new vdf hw software?

## Data 

Here is the data we gathered for all the various proofs.

### Overview

- amount of witness_types per vdf proof type
- overall size per witness_type
- currently compactifiable size

witness_type | size | rc_eos | rc_ip | rc_sp | cc_eos | cc_ip | cc_sp | icc_eos | icc_ip | size current | size optimal | size current compactifiable | size optimal compactifiable
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
0 | 100 | 550 | 18631 | 18157 | 115785 | 2901115 | 2924074 | 115896 | 17199 | 611140700 | 611140700 | 605687000 | 605687000
1 | 241 | 1391 | 38277 | 44772 | 1205 | 37920 | 44248 | 1194 | 43234 | 51150081 | 21224100 | 20380647 | 8456700
2 | 382 | 137864 | 4426425 | 4370981 | 23307 | 1546096 | 1467114 | 26906 | 4218019 | 6194783984 | 1621671200 | 1170227586 | 306342300
3 | 523 | 819 | 32572 | 24211 | 706 | 32250 | 23941 | 0 | 100 | 59935277 | 11459900 | 29757131 | 5689700
4 | 664 | 938 | 29674 | 30207 | 830 | 29337 | 29880 | 0 | 167 | 80365912 | 12103300 | 39871208 | 6004700
5 | 805 | 432 | 19420 | 13947 | 372 | 19133 | 13722 | 0 | 182 | 54102440 | 6720800 | 26747735 | 3322700
6 | 946 | 533 | 18020 | 17505 | 480 | 17717 | 17292 | 0 | 189 | 67862256 | 7173600 | 33572594 | 3548900
7 | 1087 | 254 | 11223 | 8058 | 225 | 11025 | 7944 | 0 | 116 | 42224515 | 3884500 | 20863878 | 1919400
8 | 1228 | 320 | 10518 | 10170 | 291 | 10370 | 10046 | 0 | 79 | 51323032 | 4179400 | 25428196 | 2070700
9 | 1369 | 150 | 6392 | 4573 | 135 | 6326 | 4507 | 0 | 34 | 30278173 | 2211700 | 15015192 | 1096800
10 | 1510 | 203 | 6226 | 5921 | 180 | 6191 | 5868 | 0 | 11 | 37146000 | 2460000 | 18480890 | 1223900
11 | 1651 | 79 | 3758 | 2714 | 70 | 3728 | 2678 | 0 | 2 | 21510879 | 1302900 | 10691876 | 647600
12 | 1792 | 128 | 3502 | 3596 | 112 | 3476 | 3564 | 0 | 1 | 25767168 | 1437900 | 12816384 | 715200
13 | 1933 | 60 | 2226 | 1570 | 50 | 2206 | 1554 | 0 | 0 | 14818378 | 766600 | 7364730 | 381000
14 | 2074 | 76 | 2143 | 2078 | 68 | 2129 | 2055 | 0 | 0 | 17730626 | 854900 | 8818648 | 425200
15 | 2215 | 43 | 1357 | 884 | 35 | 1344 | 875 | 0 | 0 | 10051670 | 453800 | 4992610 | 225400
16 | 2356 | 44 | 1256 | 1266 | 41 | 1255 | 1252 | 0 | 0 | 12048584 | 511400 | 6003088 | 254800
17 | 2497 | 25 | 780 | 556 | 22 | 768 | 551 | 0 | 0 | 6746894 | 270200 | 3348477 | 134100
18 | 2638 | 24 | 742 | 762 | 21 | 736 | 750 | 0 | 0 | 8006330 | 303500 | 3975466 | 150700
19 | 2779 | 10 | 440 | 347 | 8 | 436 | 344 | 0 | 0 | 4404715 | 158500 | 2189852 | 78800
20 | 2920 | 14 | 424 | 467 | 14 | 423 | 466 | 0 | 0 | 5279360 | 180800 | 2636760 | 90300
21 | 3061 | 4 | 272 | 196 | 3 | 271 | 194 | 0 | 0 | 2877340 | 94000 | 1432548 | 46800
22 | 3202 | 6 | 236 | 241 | 5 | 235 | 239 | 0 | 0 | 3080324 | 96200 | 1533758 | 47900
23 | 3343 | 8 | 166 | 121 | 8 | 165 | 119 | 0 | 0 | 1962341 | 58700 | 976156 | 29200
24 | 3484 | 5 | 158 | 147 | 4 | 155 | 145 | 0 | 0 | 2139176 | 61400 | 1059136 | 30400
25 | 3625 | 0 | 97 | 99 | 0 | 97 | 98 | 0 | 0 | 1417375 | 39100 | 706875 | 19500
26 | 3766 | 2 | 112 | 90 | 1 | 111 | 89 | 0 | 0 | 1525230 | 40500 | 756966 | 20100
27 | 3907 | 4 | 70 | 47 | 4 | 70 | 46 | 0 | 0 | 941587 | 24100 | 468840 | 12000
28 | 4048 | 2 | 61 | 46 | 2 | 60 | 45 | 0 | 0 | 874368 | 21600 | 433136 | 10700
29 | 4189 | 3 | 29 | 28 | 2 | 29 | 28 | 0 | 0 | 498491 | 11900 | 247151 | 5900
30 | 4330 | 1 | 26 | 32 | 1 | 26 | 32 | 0 | 0 | 510940 | 11800 | 255470 | 5900
31 | 4471 | 0 | 22 | 9 | 0 | 22 | 9 | 0 | 0 | 277202 | 6200 | 138601 | 3100
32 | 4612 | 2 | 23 | 29 | 2 | 23 | 28 | 0 | 0 | 493484 | 10700 | 244436 | 5300
33 | 4753 | 2 | 13 | 11 | 1 | 13 | 11 | 0 | 0 | 242403 | 5100 | 118825 | 2500
34 | 4894 | 0 | 4 | 13 | 0 | 4 | 13 | 0 | 0 | 166396 | 3400 | 83198 | 1700
35 | 5035 | 0 | 10 | 9 | 0 | 10 | 9 | 0 | 0 | 191330 | 3800 | 95665 | 1900
36 | 5176 | 0 | 10 | 8 | 0 | 10 | 8 | 0 | 0 | 186336 | 3600 | 93168 | 1800
37 | 5317 | 0 | 8 | 7 | 0 | 8 | 7 | 0 | 0 | 159510 | 3000 | 79755 | 1500
38 | 5458 | 0 | 5 | 3 | 0 | 5 | 3 | 0 | 0 | 87328 | 1600 | 43664 | 800
39 | 5599 | 0 | 4 | 3 | 0 | 4 | 3 | 0 | 0 | 78386 | 1400 | 39193 | 700
40 | 5740 | 1 | 1 | 3 | 1 | 1 | 3 | 0 | 0 | 57400 | 1000 | 28700 | 500
41 | 5881 | 0 | 2 | 3 | 0 | 2 | 3 | 0 | 0 | 58810 | 1000 | 29405 | 500
42 | 6022 | 0 | 2 | 1 | 0 | 2 | 1 | 0 | 0 | 36132 | 600 | 18066 | 300
43 | 6163 | 0 | 1 | 1 | 0 | 1 | 1 | 0 | 0 | 24652 | 400 | 12326 | 200
44 | 6304 | 0 | 2 | 2 | 0 | 2 | 2 | 0 | 0 | 50432 | 800 | 25216 | 400
45 | 6445 | 0 | 1 |   | 0 | 1 | 0 | 0 | 0 | 12890 | 200 | 6445 | 100
  |   |   |   |   |   |   |   |   |   | 6813486137 | 1699831100 | 1472109647 | 343028600
  |   |   |   |   |   |   |   |   |   |   |   |   |  
  |   |   |   |   |   |   |   |   | difference | 4.76 | GiB | 1.05 | GiB

### rc_ip proof Details

Here is a table showing the data gathered for the rc_ip proof, where the ASIC cutoff was block 4387596
witness_type | size Per proof | rc_ip before ASIC | size per witness type | Share % | rc_ip after ASIC | size per witness type | Share %
-- | -- | -- | -- | -- | -- | -- | --
0 | 100 | 17 | 1700 | 0.000387 | 18615 | 1861500 | 7.512743
1 | 241 | 24 | 5784 | 0.000547 | 38255 | 9219455 | 15.439162
2 | 382 | 4386592 | 1675678144 | 99.977117 | 39836 | 15217352 | 16.077230
3 | 523 | 108 | 56484 | 0.002461 | 32470 | 16981810 | 13.104420
4 | 664 | 171 | 113544 | 0.003897 | 29507 | 19592648 | 11.908596
5 | 805 | 201 | 161805 | 0.004581 | 19221 | 15472905 | 7.757316
6 | 946 | 204 | 192984 | 0.004649 | 17822 | 16859612 | 7.192700
7 | 1087 | 137 | 148919 | 0.003122 | 11088 | 12052656 | 4.474956
8 | 1228 | 89 | 109292 | 0.002028 | 10432 | 12810496 | 4.210203
9 | 1369 | 36 | 49284 | 0.000820 | 6358 | 8704102 | 2.565996
10 | 1510 | 11 | 16610 | 0.000251 | 6216 | 9386160 | 2.508687
11 | 1651 | 5 | 8255 | 0.000114 | 3753 | 6196203 | 1.514656
12 | 1792 | 1 | 1792 | 0.000023 | 3502 | 6275584 | 1.413356
13 | 1933 |   |   | 0.000000 | 2226 | 4302858 | 0.898381
14 | 2074 |   |   | 0.000000 | 2143 | 4444582 | 0.864884
15 | 2215 |   |   | 0.000000 | 1358 | 3007970 | 0.548069
16 | 2356 |   |   | 0.000000 | 1256 | 2959136 | 0.506903
17 | 2497 |   |   | 0.000000 | 780 | 1947660 | 0.314797
18 | 2638 |   |   | 0.000000 | 742 | 1957396 | 0.299460
19 | 2779 |   |   | 0.000000 | 440 | 1222760 | 0.177578
20 | 2920 |   |   | 0.000000 | 424 | 1238080 | 0.171120
21 | 3061 |   |   | 0.000000 | 272 | 832592 | 0.109775
22 | 3202 |   |   | 0.000000 | 236 | 755672 | 0.095246
23 | 3343 |   |   | 0.000000 | 166 | 554938 | 0.066995
24 | 3484 |   |   | 0.000000 | 158 | 550472 | 0.063767
25 | 3625 |   |   | 0.000000 | 97 | 351625 | 0.039148
26 | 3766 |   |   | 0.000000 | 112 | 421792 | 0.045202
27 | 3907 |   |   | 0.000000 | 70 | 273490 | 0.028251
28 | 4048 |   |   | 0.000000 | 61 | 246928 | 0.024619
29 | 4189 |   |   | 0.000000 | 29 | 121481 | 0.011704
30 | 4330 |   |   | 0.000000 | 26 | 112580 | 0.010493
31 | 4471 |   |   | 0.000000 | 22 | 98362 | 0.008879
32 | 4612 |   |   | 0.000000 | 23 | 106076 | 0.009282
33 | 4753 |   |   | 0.000000 | 13 | 61789 | 0.005247
34 | 4894 |   |   | 0.000000 | 4 | 19576 | 0.001614
35 | 5035 |   |   | 0.000000 | 10 | 50350 | 0.004036
36 | 5176 |   |   | 0.000000 | 10 | 51760 | 0.004036
37 | 5317 |   |   | 0.000000 | 8 | 42536 | 0.003229
38 | 5458 |   |   | 0.000000 | 5 | 27290 | 0.002018
39 | 5599 |   |   | 0.000000 | 4 | 22396 | 0.001614
40 | 5740 |   |   | 0.000000 | 1 | 5740 | 0.000404
41 | 5881 |   |   | 0.000000 | 2 | 11762 | 0.000807
42 | 6022 |   |   | 0.000000 | 2 | 12044 | 0.000807
43 | 6163 |   |   | 0.000000 | 1 | 6163 | 0.000404
44 | 6304 |   |   | 0.000000 | 2 | 12608 | 0.000807
45 | 6445 |   |   | 0.000000 | 1 | 6445 | 0.000404
  |   |   |   |   |   |   |  
  | amount proofs | 4387596 |   |   | 247779 |   |  
  | size avg per proof |   | 382.1 |   |   | 712.2 |

**Comment by wjblanke** [comment]
TLDR this is normal for ASIC proofs

It's not possible for the ASIC i/o and host to keep up with the IPS of the ASIC, so what the ASIC does is race ahead doing squaring and in the process sends back intermediate values to the host over the USB interface. 

This allows the serial squaring operation to be parallelized by the host software VDF, since it now has intermediate values from the "future" to use. This is why you still need a powerful host to run an ASIC timelord.

Unfortunately this makes determining the segments to use for the proof a lot more complicated. witness_type is a tradeoff, the higher the witness_type the easier it is to make the last proof segment super small which means the proof latency is short and the overall IPS faster. The downside is then you get a larger proof because of more segments. So the ASIC is leveraging the witness_type in order to try to calculate the proof as fast as possible.

But, this is why we have blueboxes. Blueboxes can clean up afterwards and recalculate the proof to be small because they aren't concerned about latency. The only issue now is we need 3x the blueboxes because the ASIC is 3x faster.

Thanks for the detailed data btw it is appreciated

**Comment by neurosis69** [comment]
Thanks for the detailed answer :)

> Thanks for the detailed data btw it is appreciated
Happy to help if similar data listings are needed.
